File: victor/react-native/modules/elpian-widgets/android/src/main/java/expo/modules/elpianwidgets/ElpianWidgetsModule.kt
package expo.modules.elpianwidgets

import android.util.Log
import com.facebook.react.bridge.ReactContext
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

// Installs the JSI binding global.__ElpianWidgets (see cpp/ElpianWidgetsJsi.cpp),
// which NativeWidgetRenderer drives, and registers the VictorSurfaceView that
// hosts the VM-driven native widget tree. install() returns a status string.
class ElpianWidgetsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ElpianWidgets")

    OnCreate { tryInstall() }

    Function("install") { tryInstall() }

    View(VictorSurfaceView::class) {
      // No props: the VM drives the tree through the op stream, not view props.
    }
  }

  private fun tryInstall(): String {
    try {
      System.loadLibrary("elpianwidgets_jsi")
    } catch (t: Throwable) {
      Log.e(TAG, "loadLibrary failed", t)
      return "loadLibrary: ${t.message}"
    }
    val reactContext = appContext.reactContext as? ReactContext ?: return "no-react-context"
    val jsiPtr = reactContext.javaScriptContextHolder?.get() ?: 0L
    if (jsiPtr == 0L) return "no-runtime-pointer(bridgeless=${reactContext.isBridgeless})"
    return try {
      nativeInstall(jsiPtr)
      "ok"
    } catch (t: Throwable) {
      Log.e(TAG, "nativeInstall failed", t)
      "nativeInstall: ${t.message}"
    }
  }

  private external fun nativeInstall(jsiPtr: Long)

  companion object {
    private const val TAG = "ElpianWidgetsModule"
  }
}

File: victor/react-native/modules/elpian-widgets/ios/ElpianWidgetsModule.swift
import ExpoModulesCore

// Installs the JSI binding global.__ElpianWidgets (shared cpp/ElpianWidgetsJsi.cpp),
// which NativeWidgetRenderer drives, and registers VictorSurfaceView that hosts
// the VM-driven native UIView tree. install() returns a status string. iOS twin
// of ElpianWidgetsModule.kt.
public class ElpianWidgetsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ElpianWidgets")

    OnCreate { _ = self.tryInstall() }

    Function("install") { return self.tryInstall() }

    View(VictorSurfaceView.self) {
      // The only prop: which mini-app scope this surface hosts. The VM drives the
      // tree through the op stream; appId just routes that stream to this surface.
      Prop("appId") { (view: VictorSurfaceView, appId: String) in
        view.appId = appId
      }
    }
  }

  private func tryInstall() -> String {
    // See ElpianRnModule.swift: `appContext?.runtime?.pointer` is the Expo SDK 52
    // accessor for the jsi::Runtime; verify this one line on a Mac build.
    guard let runtime = appContext?.runtime else { return "no-runtime-pointer" }
    return ElpianWidgetsBridge.install(withRuntimePointer: UInt(bitPattern: runtime.pointer))
  }
}
